File: advent_of_code/2022/day_05/solution.py
import re

from utils.py_utils.constants import INPUT
from utils.py_utils.decorators import strip_newlines, to_list


@to_list
@strip_newlines
def get_data(variation):
    with open(f"inputs/05_{variation}.txt") as f:
        data = f.readlines()
    return data


def modify(data):
    def build_start_state(levels):
        levels = [crate.replace("    ", " [-] ") for crate in levels[:-1]]
        levels = [[crate[1] for crate in level.split()] for level in levels]
        stacks = [[] for _ in range(len(levels[-1])+1)]

        for level in reversed(levels):
            for index, crate in enumerate(level):
                if crate != "-":
                    stacks[index+1].append(crate)

        return stacks

    separator = data.index("")
    levels, instructions = data[:separator], data[separator+1:]
    stacks = build_start_state(levels)

    return stacks, instructions


def easy(data):
    stacks, instructions = modify(data)
    for instruction in instructions:
        matches = re.match("move (\d+) from (\d+) to (\d+)", instruction)
        qty, frm, to = map(int, matches.groups())

        for _ in range(qty):
            crate = stacks[frm].pop()
            stacks[to].append(crate)

    return "".join(map(lambda stack: stack[-1], stacks[1:]))


def hard(data):
    stacks, instructions = modify(data)
    for instruction in instructions:
        matches = re.match("move (\d+) from (\d+) to (\d+)", instruction)
        qty, frm, to = map(int, matches.groups())

        creates_to_move = stacks[frm][-qty:]
        stacks[frm] = stacks[frm][:-qty]
        stacks[to] += creates_to_move

    return "".join(map(lambda stack: stack[-1], stacks[1:]))


if __name__ == "__main__":
    data = get_data(INPUT)
    print(easy(data))
    print(hard(data))

File: advent_of_code/2022/day_05/solution.rb
require_relative "../utils/ruby_utils/constants"


def raw_data(variation)
  File.open("inputs/05_#{variation}.txt").read
end

def modify(data)
  data = data.split(NEW_LINE)
  separator = data.find_index("")
  levels, instructions = data[..separator-2], data[separator+1..]
  levels = levels.map { |line| line.gsub("    ", " [-] ") }
                 .map { |line| line.scan(/[A-Z-]/) }

  return levels, instructions
end

def get_data(variation)
  levels, instructions = modify(raw_data(variation))
  stacks = build_start_state(levels)
  instructions = parse_instructions(instructions)

  return stacks, instructions
end

def build_start_state(levels)
  stacks = Array.new(levels[-1].size+1){ [] }
  levels.reverse.each do |level|
    level.each.with_index(1) do |crate, index|
      unless crate == "-"
        stacks[index].append(crate)
      end
    end
  end

  stacks
end

def parse_instructions(instructions)
  instructions.map do |instruction|
    instruction.match(/move (\d+) from (\d+) to (\d+)/)
               .captures
               .map(&:to_i)
  end
end

def easy(data)
  stacks, instructions = *data

  instructions.each do |qty, from, to|
    qty.times do
      crate = stacks[from].pop
      stacks[to].push(crate)
    end
  end

  stacks.map { |stack| stack[-1] }.join
end

def hard(data)
  stacks, instructions = *data

  instructions.each do |qty, from, to|
    crates_to_move = stacks[from].pop(qty)
    stacks[to].push(*crates_to_move)
  end

  stacks.map { |stack| stack[-1] }.join
end

if __FILE__ == $0
  data = get_data(INPUT)
  puts easy(data)
  puts hard(data)
end
